src/dataflow/deadlines.rs

```diff
@@ -0,0 +1,123 @@
+use crate::dataflow::{stream::StreamId, time::Timestamp, State};
+use std::{collections::HashMap, sync::Arc};
+use tokio::time::Duration;
+
+/// `CondFn` defines the type of the start and end condition functions.
+/// Each function receives the `ConditionContext` that contains the information necessary for the
+/// evaluation of the start and end conditions (s.a. the message counts and the state of the
+/// watermarks on each stream for the given timestamp)
+type CondFn = dyn Fn(&ConditionContext) -> bool + Send + Sync;
+
+/// Enumerates the types of deadlines available to the user.
+pub enum Deadline {
+    TimestampDeadline(Box<TimestampDeadline>),
+}
+
+#[derive(Debug)]
+pub struct ConditionContext {
+    message_count: HashMap<(StreamId, Timestamp), usize>,
+    watermark_status: HashMap<(StreamId, Timestamp), bool>,
```

Comment:
There is a `clear_state` method that's invoked at the end of either the completion of an end condition or at the invocation of a deadline handler. 

Let's keep the `watermark_status` as it is for now in case we need to do the `atomic usize` trick for performance later.